dm/master/server_test.go

```diff
@@ -1141,3 +1141,86 @@ func mockRevelantWorkerClient(mockWorkerClient *pbmock.MockWorkerClient, taskNam
 		},
 	).Return(queryResp, nil).MaxTimes(maxRetryNum)
 }
+
+func (t *testMaster) TestGetLeader(c *check.C) {
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
+	defer cancel()
+	// create a new cluster
+	cfg1 := NewConfig()
+	c.Assert(cfg1.Parse([]string{"-config=./dm-master.toml"}), check.IsNil)
+	cfg1.Name = "dm-master-1"
+	cfg1.DataDir = c.MkDir()
+	cfg1.MasterAddr = tempurl.Alloc()[len("http://"):]
+	cfg1.PeerUrls = tempurl.Alloc()
+	cfg1.AdvertisePeerUrls = cfg1.PeerUrls
+	cfg1.InitialCluster = fmt.Sprintf("%s=%s", cfg1.Name, cfg1.AdvertisePeerUrls)
+
+	s1 := NewServer(cfg1)
+	c.Assert(s1.Start(ctx), check.IsNil)
+
+	// wait the first one become the leader
+	c.Assert(utils.WaitSomething(30, 100*time.Millisecond, func() bool {
+		return s1.election.IsLeader()
+	}), check.IsTrue)
+
+	{
+		res, err := s1.GetLeader(ctx, &pb.GetLeaderRequest{})
+		c.Assert(err, check.IsNil)
+		c.Assert(res.Name, check.Equals, cfg1.Name)
+		c.Assert(res.Addr, check.Equals, cfg1.AdvertiseAddr)
+	}
+
+	// join to an existing cluster
+	cfg2 := NewConfig()
+	c.Assert(cfg2.Parse([]string{"-config=./dm-master.toml"}), check.IsNil)
+	cfg2.Name = "dm-master-2"
+	cfg2.DataDir = c.MkDir()
+	cfg2.MasterAddr = tempurl.Alloc()[len("http://"):]
+	cfg2.PeerUrls = tempurl.Alloc()
+	cfg2.AdvertisePeerUrls = cfg2.PeerUrls
+	cfg2.Join = cfg1.MasterAddr // join to an existing cluster
+
+	s2 := NewServer(cfg2)
+	c.Assert(s2.Start(ctx), check.IsNil)
+	defer s2.Close()
+
+	{
+		res, err := s1.GetLeader(ctx, &pb.GetLeaderRequest{})
+		c.Assert(err, check.IsNil)
+		c.Assert(res.Name, check.Equals, cfg1.Name)
+		c.Assert(res.Addr, check.Equals, cfg1.AdvertiseAddr)
+	}
+	{
+		res, err := s2.GetLeader(ctx, &pb.GetLeaderRequest{})
+		c.Assert(err, check.IsNil)
+		c.Assert(res.Name, check.Equals, cfg1.Name)
+		c.Assert(res.Addr, check.Equals, cfg1.AdvertiseAddr)
+	}
+
+	s1.Close()
+	{
+		res, err := s2.GetLeader(ctx, &pb.GetLeaderRequest{})
+		c.Assert(err, check.IsNil)
+		c.Assert(res.Name, check.Equals, "")
+		c.Assert(res.Addr, check.Equals, "")
+	}
+
+	/*
+		It seems that if we kill leader in 2 masters cluster,
+		the second one wouldn't be the leader so follow cases would not be passed.
+		But I fail to create 3 masters cluster...
```

Comment:
what is the error log of creating the third master?

Reply:
The problem has been solved by waiting for the second master start.